core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala

```diff
@@ -356,4 +374,233 @@ private[mesos] trait MesosSchedulerUtils extends Logging {
+    val nonZeroPorts = portsToCheck.filter(_ != 0)
+    val withinRange = nonZeroPorts.forall(p => checkIfInRange(p, ports))
+    // make sure we have enough ports to allocate per offer
+    ports.map(r => r._2 - r._1 + 1).sum >= portsToCheck.size && withinRange
+  }
+
+  /**
+   * Partitions port resources.
+   *
+   * @param conf the spark config
+   * @param ports the ports offered
+   * @return resources left, port resources to be used and the list of assigned ports
+   */
+  def partitionPorts(
+      conf: SparkConf,
+      ports: List[Resource])
+    : (List[Resource], List[Resource], List[Long]) = {
+    val taskPortRanges = getRangeResourceWithRoleInfo(ports.asJava, "ports")
+    val portsToCheck = ManagedPorts.getPortValues(conf)
+    val nonZeroPorts = portsToCheck.filter(_ != 0)
+    // reserve non zero ports first
+    val nonZeroResources = reservePorts(taskPortRanges, nonZeroPorts)
+    // reserve actual port numbers for zero ports - not set by the user
+    val numOfZeroPorts = portsToCheck.count(_ == 0)
+    val randPorts = pickRandomPortsFromRanges(nonZeroResources._1, numOfZeroPorts)
+    val zeroResources = reservePorts(nonZeroResources._1, randPorts)
+    val (portResourcesLeft, portResourcesToBeUsed) =
+      createResources(nonZeroResources, zeroResources)
+    (portResourcesLeft, portResourcesToBeUsed, nonZeroPorts ++ randPorts)
+  }
+
+  private object ManagedPorts {
+    val portNames = List("spark.executor.port", "spark.blockManager.port")
+
+    def getPortValues(conf: SparkConf): List[Long] = {
+      portNames.map(conf.getLong(_, 0))
+    }
+  }
+
+  private def createResources(
+      nonZero: (List[PortRangeResourceInfo], List[PortRangeResourceInfo]),
+      zero: (List[PortRangeResourceInfo], List[PortRangeResourceInfo]))
+    : (List[Resource], List[Resource]) = {
+    val resources = {
+      if (nonZero._2.isEmpty) { // no user ports were defined
+        (zero._1.flatMap{port => createMesosPortResource(port.range, Some(port.role))},
+          zero._2.flatMap{port => createMesosPortResource(port.range, Some(port.role))})
+
+      } else if (zero._2.isEmpty) { // no random ports were defined
+        (nonZero._1.flatMap{port => createMesosPortResource(port.range, Some(port.role))},
+          nonZero._2.flatMap{port => createMesosPortResource(port.range, Some(port.role))})
+      }
+      else {  // we have user defined and random ports defined
+        val left = zero._1.flatMap{port => createMesosPortResource(port.range, Some(port.role))}
+        val used = nonZero._2.flatMap{port =>
+          createMesosPortResource(port.range, Some(port.role))} ++
+          zero._2.flatMap{port => createMesosPortResource(port.range, Some(port.role))}
+        (left, used)
+      }
+    }
+    resources
+  }
+
+  private case class PortRangeResourceInfo(role: String, range: List[(Long, Long)])
+
+  private def getRangeResourceWithRoleInfo(res: JList[Resource], name: String)
```

Comment:
`res` is singular, but this is a list.  call this `resources`.  it's also the name we use elsewhere